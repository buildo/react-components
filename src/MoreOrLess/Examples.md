### Examples

```js
initialState = { expanded: false };

function onExpandedChange() {
  setState({ expanded: !state.expanded });
}

const templateExpanded = () => (
  <FlexView className="card">
    <FlexView grow shrink className="description">
      <FlexView column>
        <FlexView className="title">The Student Hotel</FlexView>
        <FlexView className="subtitle">
          <FlexView className="author">Taylor Cambell</FlexView>
          <FlexView className="date">7th of March</FlexView>
        </FlexView>
        What is boutique hotel design, and why does it matter? There are fashion trends in hotel
        design just as there are in other areas. In the late eighties and through the nineties, that
        fashion was for clean, stark, minimalist design, particularly in city hotels. This was such
        a diversion from the traditional way.
      </FlexView>
    </FlexView>
    <FlexView grow shrink={false} hAlignContent="right">
      <img src="moreOrLess.png" />
    </FlexView>
  </FlexView>
);

const templateCollapsed = () => (
  <FlexView className="card" style={{ maxHeight: 116 }}>
    <FlexView className="description">
      <FlexView column>
        <FlexView className="title">The Student Hotel</FlexView>
        <FlexView className="subtitle">
          <FlexView className="author">Taylor Cambell</FlexView>
          <FlexView className="date">7th of March</FlexView>
        </FlexView>
      </FlexView>
    </FlexView>
    <FlexView grow hAlignContent="right" vAlignContent="center">
      <img src="moreOrLess2.png" style={{ width: 'auto', height: 80 }} />
    </FlexView>
  </FlexView>
);

const angleUpIcon = (
  <svg version="1.1" xmlns="http://www.w3.org/2000/svg" height="17px" viewBox="0 0 32 32">
    <path
      fill="#9098a7"
      d="M0.932 20.708l13.333-13.333c0.482-0.479 1.146-0.775 1.88-0.775s1.398 0.296 1.88 0.776l13.333 13.333c0.399 0.463 0.641 1.071 0.641 1.735 0 1.473-1.194 2.667-2.667 2.667-0.664 0-1.271-0.243-1.738-0.644l-13.33-13.33h3.76l-13.333 13.333c-0.492 0.572-1.217 0.932-2.025 0.932-1.473 0-2.667-1.194-2.667-2.667 0-0.809 0.36-1.533 0.929-2.023l0.003-0.003z"
    ></path>
  </svg>
);

const angleDownIcon = (
  <svg version="1.1" xmlns="http://www.w3.org/2000/svg" height="17px" viewBox="0 0 32 32">
    <path
      fill="#9098a7"
      d="M0.647 11.102l13.456 13.456c0.486 0.484 1.157 0.783 1.897 0.783s1.411-0.299 1.897-0.783l13.456-13.456c0.402-0.467 0.647-1.080 0.647-1.751 0-1.486-1.205-2.691-2.691-2.691-0.67 0-1.283 0.245-1.754 0.65l-13.452 13.453h3.794l-13.456-13.456c-0.468-0.402-1.080-0.647-1.75-0.647-1.486 0-2.691 1.205-2.691 2.691 0 0.67 0.245 1.283 0.65 1.754l-0.003-0.004z"
    ></path>
  </svg>
);

const { expanded } = state;
const icons = { expanded: angleUpIcon, collapsed: angleDownIcon };

<MoreOrLess {...{ expanded, onExpandedChange, icons }}>
  {expanded ? templateExpanded() : templateCollapsed()}
</MoreOrLess>;
```
