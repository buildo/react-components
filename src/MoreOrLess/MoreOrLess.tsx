import * as React from "react";
import * as cx from "classnames";
import FlexView from "react-flexview";
import { Children } from "../utils";

export namespace MoreOrLess {
  export type Props = {
    /** panel content */
    children: Children;
    /** whether the panel should be expanded or not */
    expanded: boolean;
    /** called on toggle */
    onExpandedChange: (expanded: boolean) => void;
    /** icons for expanded and collapsed panel */
    icons: {
      expanded: JSX.Element;
      collapsed: JSX.Element;
    };
    /** props for wrapper FlexView */
    wrapperProps?: object;
    /** an optional class name to pass to the component */
    className?: string;
    /** an optional style object to pass to the component */
    style?: React.CSSProperties;
  };
}

/**
 * A panel used to alternately display short or long version of the content
 */
export class MoreOrLess extends React.PureComponent<MoreOrLess.Props> {
  toggleExpanded = () => {
    this.props.onExpandedChange(!this.props.expanded);
  };

  templateExpandButton = (
    icon: JSX.Element,
    toggleExpanded: React.EventHandler<React.SyntheticEvent<HTMLDivElement>>
  ) => {
    return (
      <FlexView
        hAlignContent="center"
        vAlignContent="center"
        className="expand-button"
        onClick={toggleExpanded}
        shrink={false}
      >
        {icon}
      </FlexView>
    );
  };

  render() {
    const {
      props: { children, className, style, expanded, icons, wrapperProps },
      toggleExpanded
    } = this;

    const panelState = expanded ? "more" : "less";
    const icon = expanded ? icons.expanded : icons.collapsed;

    return (
      <FlexView
        column
        shrink={false}
        {...wrapperProps}
        style={style}
        className={cx("more-or-less", panelState, className)}
      >
        {children}
        {this.templateExpandButton(icon, toggleExpanded)}
      </FlexView>
    );
  }
}
