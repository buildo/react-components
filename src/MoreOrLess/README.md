A panel used to alternately display a short (less) or a long (more) version of the content.

![less](https://user-images.githubusercontent.com/10867086/36533761-7b550daa-17bc-11e8-9be2-a406e1716adf.jpg)

![more](https://user-images.githubusercontent.com/10867086/36533806-994e8ff2-17bc-11e8-8e59-2f5e721b7b2f.jpg)

The short content should be a summary of the long content. Remember the following:
- do not arbitrarily cut the content in the short version, try to provide a summary with the most relevant information
- it's ok to change the layout between the short and long version, to optimize how information is presented
- this is not an overflow component, the content should be semantically the same but summarized in the short version, with only the most relevant information presented

