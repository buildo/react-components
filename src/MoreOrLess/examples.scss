@import '../utils/theme/variables.scss';

.more-or-less {
  border: 1px solid $brc-silver;

  .card {
    padding: 25px 33px;

    .description {
      margin-right: 20px;
      font-size: 14px;
      line-height: 1.5;

      .title {
        font-size: 28px;
        font-weight: $brc-bold;
      }

      .subtitle {
        padding-top: 10px;
        padding-bottom: 15px;
        font-size: 14px;

        .author {
          padding-right: 9px;
          border-right: 1px solid $brc-silver;
          color: $brc-azure;

          &::before {
            content: 'by\00a0';
            color: $brc-darkGreyBlue;
          }
        }

        .date {
          padding-left: 9px;
        }
      }
    }
  }

  .expand-button {
    .icon {
      font-size: 8px;
    }
  }
}
