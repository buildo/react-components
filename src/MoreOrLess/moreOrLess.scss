@import '../utils/theme/variables.scss';
$expand-button-background: $brc-gradientGrey !default;
$expand-button-background-hover: $brc-gradientDarkGrey !default;
$expand-button-icon-color: $brc-coolGrey !default;
$expand-button-icon-color-hover: $brc-coolGrey !default;

.more-or-less {
  .expand-button {
    border-top: 1px solid $brc-silver;
    height: 15px;
    width: 100%;
    background: $expand-button-background;
    cursor: pointer;

    &:hover {
      background-color: $expand-button-background-hover;
    }
  }
}
