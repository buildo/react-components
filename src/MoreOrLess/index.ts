import { MoreOrLess } from './MoreOrLess';
export default MoreOrLess;
