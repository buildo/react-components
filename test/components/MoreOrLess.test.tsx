import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { mount, shallow } from 'enzyme';

import MoreOrLess from '../../src/MoreOrLess';

let consoleError: jest.SpyInstance<{}>;

beforeAll(() => {
  consoleError = jest.spyOn(console, 'error');
});

afterEach(() => {
  expect(consoleError).not.toHaveBeenCalled();
});

const exampleProps = {
  children: 'content',
  className: 'fancy-class-name',
  icons: {
    expanded: 'angle-up',
    collapsed: 'angle-down'
  },
  onExpandedChange: () => {}
};

const componentMore = mount(<MoreOrLess {...exampleProps} expanded={true} />);

const componentLess = mount(<MoreOrLess {...exampleProps} expanded={false} />);

describe('MoreOrLess', () => {
  it('renders correctly when expanded', () => {
    const component = renderer.create(<MoreOrLess {...exampleProps} expanded />);
    expect(component).toMatchSnapshot();
  });

  it('renders correctly when not expanded', () => {
    const component = renderer.create(<MoreOrLess {...exampleProps} expanded={false} />);
    expect(component).toMatchSnapshot();
  });
});
