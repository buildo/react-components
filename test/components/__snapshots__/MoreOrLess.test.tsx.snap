// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoreOrLess renders correctly when expanded 1`] = `
<div
  className="more-or-less more fancy-class-name"
  style={
    Object {
      "alignItems": undefined,
      "boxSizing": "border-box",
      "display": "flex",
      "flex": "0 0 auto",
      "flexDirection": "column",
      "flexWrap": "nowrap",
      "height": undefined,
      "justifyContent": undefined,
      "marginBottom": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "minHeight": 0,
      "minWidth": 0,
      "width": undefined,
    }
  }
>
  content
  <div
    className="expand-button"
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "boxSizing": "border-box",
        "display": "flex",
        "flex": "0 0 auto",
        "flexDirection": "row",
        "flexWrap": "nowrap",
        "height": undefined,
        "justifyContent": "center",
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "minHeight": 0,
        "minWidth": 0,
        "width": undefined,
      }
    }
  >
    angle-up
  </div>
</div>
`;

exports[`MoreOrLess renders correctly when not expanded 1`] = `
<div
  className="more-or-less less fancy-class-name"
  style={
    Object {
      "alignItems": undefined,
      "boxSizing": "border-box",
      "display": "flex",
      "flex": "0 0 auto",
      "flexDirection": "column",
      "flexWrap": "nowrap",
      "height": undefined,
      "justifyContent": undefined,
      "marginBottom": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "minHeight": 0,
      "minWidth": 0,
      "width": undefined,
    }
  }
>
  content
  <div
    className="expand-button"
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "boxSizing": "border-box",
        "display": "flex",
        "flex": "0 0 auto",
        "flexDirection": "row",
        "flexWrap": "nowrap",
        "height": undefined,
        "justifyContent": "center",
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "minHeight": 0,
        "minWidth": 0,
        "width": undefined,
      }
    }
  >
    angle-down
  </div>
</div>
`;
